mkosi

```diff
@@ -79,6 +79,73 @@ MKOSI_COMMANDS_SUDO = ("build", "clean") + MKOSI_COMMANDS_CMDLINE
 MKOSI_COMMANDS = ("build", "clean", "help", "summary") + MKOSI_COMMANDS_CMDLINE
 
 
+DRACUT_UNIFIED_KERNEL_INSTALL = """\
+#!/bin/bash -e
+
+COMMAND="$1"
+KERNEL_VERSION="$2"
+BOOT_DIR_ABS="$3"
+KERNEL_IMAGE="$4"
+ROOTHASH="${5:-}"
+
+# If KERNEL_INSTALL_MACHINE_ID is defined but empty, BOOT_DIR_ABS is a fake directory so let's skip creating
+# the unified kernel image.
+if [ -z "${KERNEL_INSTALL_MACHINE_ID-unset}" ]; then
+    exit 0
+fi
+
+# Strip machine ID and kernel version to get the boot directory.
+PREFIX=$(dirname $(dirname "$BOOT_DIR_ABS"))
+
+if [[ -n "$ROOTHASH" ]]; then
+    BOOT_BINARY="${PREFIX}/EFI/Linux/linux-${KERNEL_VERSION}-${ROOTHASH}.efi"
+else
+    BOOT_BINARY="${PREFIX}/EFI/Linux/linux-${KERNEL_VERSION}.efi"
+fi
+
+case "$COMMAND" in
+    add)
+        if [[ -f /etc/kernel/cmdline ]]; then
+            read -r -d '' BOOT_OPTIONS < /etc/kernel/cmdline || true
+        elif [[ -f /usr/lib/kernel/cmdline ]]; then
+            read -r -d '' BOOT_OPTIONS < /usr/lib/kernel/cmdline || true
+        else
+            read -r -d '' BOOT_OPTIONS < /proc/cmdline || true
+        fi
+
+        if [[ -n "$ROOTHASH" ]]; then
+            BOOT_OPTIONS="${BOOT_OPTIONS} roothash=$ROOTHASH"
+        fi
+
+        if [[ -n "$KERNEL_IMAGE" ]]; then
+            DRACUT_KERNEL_IMAGE_OPTION="--kernel-image ${KERNEL_IMAGE}"
+        else
+            DRACUT_KERNEL_IMAGE_OPTION=""
+        fi
+
+        dracut \\
+            --quiet \\
+            --force \\
+            --no-hostonly \\
```

Comment:
Since this script replaces the normal dracut kernel-install plugins, it might make sense to also have a replacement for the one that generates the rescue image and have the `--no-hostonly` only in that one.

Reply:
mkosi is a tool for making disk images though. I don't think we should limit ourselves to the system we're currently running on since the image might be deployed on a different device.

Reply:
What we should do though is add a hostonly option to configure this. At thesame time we should also move whatever we can out of the install script and into dracut config files.

Reply:
> mkosi is a tool for making disk images though. I don't think we should limit ourselves to the system we're currently running on since the image might be deployed on a different device.

That's true, my thought, though, was that the files stay around. If those images are used longer, it would make sense to install them in generic mode (`--no-hostonly`), but have both lying around for future kernel upgrades.

Reply:
Maybe we could do image generation with --no-hostonly and the configure --hostonly afterwards so subsequent updates are faster since by then the image is likely deployed. Implementing that would be a bit awkward though. For unified kernels we'd have to pass an extra parameter to the dracut script and for normal kernel install we'd have to write a temporary dracut config file to override hostonly.